bitwarden_license/src/Scim/Commands/Users/GetUserCommand.cs

```diff
@@ -0,0 +1,27 @@
+﻿using Bit.Core.Exceptions;
+using Bit.Core.Repositories;
+using Bit.Scim.Commands.Users.Interfaces;
+using Bit.Scim.Models;
+
+namespace Bit.Scim.Commands.Users;
+
+public class GetUserCommand : IGetUserCommand
+{
+    private readonly IOrganizationUserRepository _organizationUserRepository;
+
+    public GetUserCommand(IOrganizationUserRepository organizationUserRepository)
+    {
+        _organizationUserRepository = organizationUserRepository;
+    }
+
+    public async Task<ScimUserResponseModel> GetUserAsync(Guid organizationId, Guid id)
+    {
+        var orgUser = await _organizationUserRepository.GetDetailsByIdAsync(id);
+        if (orgUser == null || orgUser.OrganizationId != organizationId)
+        {
+            throw new NotFoundException("User not found.");
```

Comment:
We return a `ScimUserResponseModel` from this query if successful, but if it fails, we throw an exception and construct the `ScimErrorResponseModel` at the controller level instead. This feels inconsistent to me, like we're exposing details to the controller level.

It also suggests a pattern where we have to wrap every query in a try/catch at the controller level to handle errors - which is going to be a lot of boilerplate across all your PRs. Can we just throw the exception in the query instead and have it automatically returned by the controller?

Reply:
That was actually my first approach - the Controller would be agnostic to the query/command behavior, it just returned whatever the result was. 

This turned out to be a little more complicated because the returned status code depends on if the query/command was successful or not. Then I also had the problem that the success and error responses were different classes (`ScimUserResponseModel` or `ScimErrorResponseModel`) so I had to create a base model to wrap them so then the Controller could identify which one it was to return. On some cases - like Users POST - the controller can return different status codes depending on the error then that meant it would be the command determining the status code that would be returned by the controller.

While the intention was to simplify the controller's actions it seemed like it was just adding an extra layer of complication.

I was more happy with this solution instead because the business logic is inside the query/command - it just returns what it is meant to and if not then it returns an exception. Also the controller only has the responsibility of setting the returned status code and creating an error response.

Reply:
How about using an [ExceptionFilterAttribute](https://learn.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.mvc.filters.exceptionfilterattribute?view=aspnetcore-6.0) on the SCIM controllers? It would let you maintain this separation, where the Command/Query throws the applicable exception, but instead of handling it in every controller method individually, you'd just handle it in a single location in the ExceptionFilterAttribute.

This is the pattern we use in the api project currently: see https://github.com/bitwarden/server/blob/0bb5baf7e5cb0b8950813c1acb3c1bc6a8a91dad/src/Api/Utilities/ExceptionHandlerFilterAttribute.cs as an example. That also shows how it can handle different exceptions differently. The really nice thing is that when you're adding new codepaths, you can just throw exceptions and know that they'll be handled appropriately.

Reply:
That is a good idea! I've implemented that on all branches.